// nextra/dist/context.js
"use strict";
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/context.ts
var context_exports = {};
__export(context_exports, {
  getAllPages: () => getAllPages,
  getCurrentLevelPages: () => getCurrentLevelPages,
  getPagesUnderRoute: () => getPagesUnderRoute
});
module.exports = __toCommonJS(context_exports);
function getContext(name) {
  const context = globalThis.__nextra_internal__;
  if (!context) {
    throw new Error(`Nextra context not found. Please make sure you are using "${name}" of "nextra/context" on a Nextra page.`);
  }
  return context;
}
function normalizeMeta(meta) {
  if (typeof meta === "string") {
    meta = {
      title: meta
    };
  }
  return meta;
}
function filter(pageMap, activeLevel) {
  var _a;
  let activeLevelPages;
  let meta = ((_a = pageMap.find((item) => item.name === "meta.json")) == null ? void 0 : _a.meta) || {};
  const metaKeys = Object.keys(meta);
  const items = [];
  for (const item of pageMap) {
    if (item.name === "meta.json")
      continue;
    const page = __spreadProps(__spreadValues({}, item), {
      meta: normalizeMeta(meta[item.name])
    });
    if (item.children) {
      const filteredChildren = filter(item.children, activeLevel);
      page.children = filteredChildren[0];
      if (filteredChildren[1]) {
        activeLevelPages = filteredChildren[1];
      } else if (page.route === activeLevel) {
        activeLevelPages = activeLevelPages || page.children;
      }
    }
    items.push(page);
  }
  return [
    items.sort((a, b) => {
      const indexA = metaKeys.indexOf(a.name);
      const indexB = metaKeys.indexOf(b.name);
      if (indexA === -1 && indexB === -1)
        return a.name < b.name ? -1 : 1;
      if (indexA === -1)
        return 1;
      if (indexB === -1)
        return -1;
      return indexA - indexB;
    }),
    activeLevelPages
  ];
}
function getAllPages() {
  const internal = getContext("getAllPages");
  return filter(internal.pageMap)[0];
}
function getCurrentLevelPages() {
  const internal = getContext("getCurrentLevelPages");
  return filter(internal.pageMap, internal.route)[1] || [];
}
function getPagesUnderRoute(route) {
  const internal = getContext("getPagesUnderRoute");
  return filter(internal.pageMap, route)[1] || [];
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  getAllPages,
  getCurrentLevelPages,
  getPagesUnderRoute
});
